README
sp-measure library provides API to make and compare snapshots of
various system and process resources.

This API can be used to gather system/process resource usage statistics - 
for example cpu usage, free memory, process memory usage etc.

See an example of API usage in examples/ directory.

changelog
sp-measure (1.3.5), 25 May 2012:
  * Disable obsolete Maemo 4 memory watermarks by default
    to avoid run-time warning.
  * Speed up SMAPS data parsing by tens of percents.
  * Document & comment typo and compiler warning fixes.
  * Fix test program CPU freq value calculation.

sp-measure (1.3.4), 23 Mar 2012:
  * Have default CFLAGS, don't default to /usr, allow doc directory
    install path to be configured.
  * Some code refactoring.

sp-measure (1.3.3), 11 Nov 2011:
  * Fix file descriptor leak in get_process_name().

sp-measure (1.3.2), 9 Sep 2011:
  * Fix CPU frequency calculation.

sp-measure (1.3.1), 5 Aug 2011:
  * Fix crashes when /proc/ reading fails.

sp-measure (1.3), 10 May 2011:
  * Add support for Cgroups memory usage monitoring

sp-measure (1.2.1), 12 Oct 2010:
  * Packaging fixes.

sp-measure (1.2), 24 Sep 2010:
  * Fix sp_measure_process.h, so that it can be included in C++ code.

sp-measure (1.1.1), 11 May 2010:
  * use LGPL as license.  

sp-measure (1.1), 30 Apr 2010:
  * API change, allow retrieving of part of the data to fail and retrieving
    of rest of data to continue. 
  * System memory usage should take swap into account. 

sp-measure (1.0), 12 Apr 2010:
  * Initial Release. 


tests/test_sp_measure.c
/*
 * This file is a spart of sp-measure library.
 *
 * Copyright (C) 2010,2012 by Nokia Corporation
 *
 * Contact: Eero Tamminen <eero.tamminen@nokia.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301  USA
 */

/**
 * Simple measure API tests.
 *
 * As most of measure API data is retrieved by parsing /proc/ files,
 * it's possible to test it by setting fake rootfs system and checking
 * if the retrived values are correct.
 * Because process snapshots stores the path to required /proc/ files
 * (currently smaps and stats), changing rootfs after snapshot initialization
 * will have no effect. This leads to a hackish approach to build the
 * second snapshot for comparison tests.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <sp_measure.h>

#define TEST(expression, ...) if (!(expression)) {fprintf(stderr, "[failure] " #expression "\n" __VA_ARGS__ ); exit(-1);}

#define TEST_VALUE_INT(expression, value)	TEST(expression == value, "\t" #expression "=%d\n", expression)
#define TEST_VALUE_STR(expression, value)	TEST(strcmp(expression, value) == 0, "\t" #expression "=%s\n", expression)

#define SNAPSHOT_TEST_SYS  (SNAPSHOT_SYS | SNAPSHOT_SYS_MEM_WATERMARK)
#define SNAPSHOT_TEST_PROC (SNAPSHOT_PROC)


void check_system_api()
{
	sp_measure_sys_data_t data1, data2, data3;

	/* set the fake rootfs */
	sp_measure_set_fs_root("./rootfs1");

	/* initialize data1, data2 snapshot structures */
	TEST(sp_measure_init_sys_data(&data1, SNAPSHOT_TEST_SYS, NULL) == 0);
	TEST(sp_measure_init_sys_data(&data2, 0, &data1) == 0);
	TEST(data1.common != NULL);
	TEST(data1.common == data2.common);

	/* check for global system parameters retrieved during initialization */
	TEST_VALUE_INT(data1.common->mem_total, 3096748);
	TEST_VALUE_INT(data1.common->mem_swap, 5111800);
	TEST_VALUE_INT(data1.common->cpu_max_freq, 2201000);

	/* take system resource usage snapshot */
	TEST(sp_measure_get_sys_data(&data1, SNAPSHOT_TEST_SYS, "snapshot1") == 0);
	TEST_VALUE_STR(data1.name, "snapshot1");

	/* check memory usage data */
	TEST_VALUE_INT(data1.mem_free, 460588);
	TEST_VALUE_INT(data1.mem_buffers, 304976);
	TEST_VALUE_INT(data1.mem_cached, 1593264);
	TEST_VALUE_INT(data1.mem_watermark, 3);

	/* check cpu usage data */
	TEST_VALUE_INT(data1.cpu_ticks_total, 85277555);
	TEST_VALUE_INT(data1.cpu_ticks_idle, 82387691);
	TEST_VALUE_INT(data1.cpu_freq_ticks_count, 5);

	/* set the fake rootfs */
	sp_measure_set_fs_root("./rootfs2");
	/* take the second snapshot */
	TEST(sp_measure_get_sys_data(&data2, SNAPSHOT_TEST_SYS, NULL) == 0);
	TEST(data2.name == NULL);
	/* take third snapshot */
	TEST(sp_measure_init_sys_data(&data3, SNAPSHOT_TEST_SYS, NULL) == 0);
	TEST(sp_measure_get_sys_data(&data3, SNAPSHOT_TEST_SYS, NULL) == 0);

	/* check memory usage data */
	TEST_VALUE_INT(data2.mem_free, 426176);
	TEST_VALUE_INT(data2.mem_buffers, 305772);
	TEST_VALUE_INT(data2.mem_cached, 1626056);
	TEST_VALUE_INT(data2.mem_watermark, 3);

	/* check cpu usage data */
	TEST_VALUE_INT(data2.cpu_ticks_total, 85580441);
	TEST_VALUE_INT(data2.cpu_ticks_idle, 82665362);
	TEST_VALUE_INT(data2.cpu_freq_ticks_count, 5);

	int diff;

	/* this isn't actually a diff, but average frequency between snapshots */
	TEST(sp_measure_diff_sys_cpu_avg_freq(&data1, &data2, &diff) == 0);
	TEST(diff == 839559, "\tsp_measure_diff_sys_cpu_avg_freq: diff=%d\n", diff);

	/* data1 and data3 does not share the same common data (they were
	 * initialized separately. Comparison operations for such snapshots
	 * should fail
	 */
	TEST(sp_measure_diff_sys_cpu_ticks(&data1, &data3, &diff) < 0);
	TEST(sp_measure_diff_sys_cpu_usage(&data1, &data3, &diff) < 0);
	TEST(sp_measure_diff_sys_cpu_avg_freq(&data1, &data3, &diff) < 0);
	TEST(sp_measure_diff_sys_mem_used(&data1, &data3, &diff) < 0);

	/* check snapshot comparison results */
	TEST(sp_measure_diff_sys_cpu_ticks(&data1, &data2, &diff) == 0);
	TEST(diff == 302886, "\tsp_measure_diff_sys_cpu_ticks: diff=%d\n", diff);

	TEST(sp_measure_diff_sys_cpu_usage(&data1, &data2, &diff) == 0);
	TEST(diff == 832, "\tsp_measure_diff_sys_cpu_usage: diff=%d\n", diff);

	TEST(sp_measure_diff_sys_mem_used(&data1, &data2, &diff) == 0);
	TEST(diff == 824, "\tsp_measure_diff_sys_mem_used: diff=%d\n", diff);

	/* reset the fake rootfs */
	sp_measure_set_fs_root(NULL);

	/* release resources allocated by snapshots */
	TEST(sp_measure_free_sys_data(&data1) == 0);
	TEST(sp_measure_free_sys_data(&data2) == 0);
	TEST(sp_measure_free_sys_data(&data3) == 0);
}


void check_process_api()
{
	sp_measure_proc_data_t data1, data2, data3;

	/* set the fake rootfs */
	sp_measure_set_fs_root("./rootfs1");
	/* initialize data1, data2 snapshot structures */
	TEST(sp_measure_init_proc_data(&data1, 25268, SNAPSHOT_TEST_PROC, NULL) == 0);
	TEST(sp_measure_init_proc_data(&data2, 25268, 0, &data1) == 0);
	TEST(data1.common != NULL);
	TEST(data1.common == data2.common);

	/* check common process parameters */
	TEST_VALUE_STR(FIELD_PROC_NAME(&data1), "eclipse");
	TEST_VALUE_INT(FIELD_PROC_PID(&data1), 25268);

	/* take process snapshot */
	TEST(sp_measure_get_proc_data(&data1, SNAPSHOT_TEST_PROC, "snapshot1") == 0);
	TEST_VALUE_STR(data1.name, "snapshot1");

	/* check memory usage data */
	TEST_VALUE_INT(FIELD_PROC_MEM_PRIVATE_CLEAN(&data1), 14104);
	TEST_VALUE_INT(FIELD_PROC_MEM_PRIVATE_DIRTY(&data1), 95992);
	TEST_VALUE_INT(FIELD_PROC_MEM_SWAP(&data1), 16192);
	TEST_VALUE_INT(FIELD_PROC_MEM_SIZE(&data1), 686500);
	TEST_VALUE_INT(FIELD_PROC_MEM_PSS(&data1), 110781);
	TEST_VALUE_INT(FIELD_PROC_MEM_RSS(&data1), 114404);
	TEST_VALUE_INT(FIELD_PROC_MEM_REFERENCED(&data1), 68956);
	TEST_VALUE_INT(FIELD_PROC_MEM_SHARED_CLEAN(&data1), 3540);
	TEST_VALUE_INT(FIELD_PROC_MEM_SHARED_DIRTY(&data1), 768);
	TEST_VALUE_INT(FIELD_PROC_MEM_PRIV_DIRTY_SUM(&data1), 112184);

	/* check cpu usage data */
	TEST_VALUE_INT(FIELD_PROC_CPU_STIME(&data1), 47282);
	TEST_VALUE_INT(FIELD_PROC_CPU_UTIME(&data1), 262287);

	/* set the fake roorfs */
	sp_measure_set_fs_root("./rootfs2");
	/* initialize and take third snapshot */
	TEST(sp_measure_init_proc_data(&data3, 25268, SNAPSHOT_TEST_PROC, NULL) == 0);
	TEST(data3.common != NULL);
	TEST(sp_measure_get_proc_data(&data3, SNAPSHOT_TEST_PROC, NULL) == 0);
	TEST(data3.name == NULL);

	/* check memory usage data */
	TEST_VALUE_INT(FIELD_PROC_MEM_PRIVATE_CLEAN(&data3), 14300);
	TEST_VALUE_INT(FIELD_PROC_MEM_PRIVATE_DIRTY(&data3), 97096);
	TEST_VALUE_INT(FIELD_PROC_MEM_SWAP(&data3), 15084);
	TEST_VALUE_INT(FIELD_PROC_MEM_SIZE(&data3), 686500);
	TEST_VALUE_INT(FIELD_PROC_MEM_PSS(&data3), 112140);
	TEST_VALUE_INT(FIELD_PROC_MEM_RSS(&data3), 116108);
	TEST_VALUE_INT(FIELD_PROC_MEM_REFERENCED(&data3), 75672);
	TEST_VALUE_INT(FIELD_PROC_MEM_SHARED_CLEAN(&data3), 3944);
	TEST_VALUE_INT(FIELD_PROC_MEM_SHARED_DIRTY(&data3), 768);
	TEST_VALUE_INT(FIELD_PROC_MEM_PRIV_DIRTY_SUM(&data3), 112180);

	/* check cpu usage data */
	TEST_VALUE_INT(FIELD_PROC_CPU_STIME(&data3), 47299);
	TEST_VALUE_INT(FIELD_PROC_CPU_UTIME(&data3), 262479);


	int diff;

	/* data1 and data3 does not share the same common data (they  wer
	 * initalized separately. Comparison operations for such snapshots
	 * should fail */
	TEST(sp_measure_diff_proc_mem_private_dirty(&data1, &data3, &diff) < 0);
	TEST(sp_measure_diff_proc_cpu_ticks(&data1, &data3, &diff) < 0);

	/* copy over data2 data to data3, restore data3.common reference.
	 * We can't compare data1 and data2 directly, as they have different common
	 * data sets. So we create data3 from the data1 and set the fields
	 * to data2 values. */
	data2 = data3;
	data2.common = data1.common;

	/* check snapshot comparison values */
	TEST(sp_measure_diff_proc_mem_private_dirty(&data1, &data2, &diff) == 0);
	TEST(diff == -4, "\tsp_measure_diff_proc_mem_private_dirty: diff=%d\n", diff);

	TEST(sp_measure_diff_proc_cpu_ticks(&data1, &data2, &diff) == 0);
	TEST(diff == 209, "\tsp_measure_diff_proc_cpu_ticks: diff=%d\n", diff);

	/* reset the fake rootfs */
	sp_measure_set_fs_root(NULL);

	/* release resources allocated by snapshots */
	TEST(sp_measure_free_proc_data(&data1) == 0);
	TEST(sp_measure_free_proc_data(&data2) == 0);
	TEST(sp_measure_free_proc_data(&data3) == 0);
}

int main() 
{
	check_system_api();
	
	check_process_api();

	return 0;
}


src/sp_measure.h
/*
 * This file is a part of sp-measure library.
 *
 * Copyright (C) 2010-2012 by Nokia Corporation
 *
 * Contact: Eero Tamminen <eero.tamminen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301  USA
 */
#ifndef SP_MEASURE_H
#define SP_MEASURE_H

#include <stdio.h>

/**
 * System resource identifiers and their generic groupings.
 */
typedef enum {
	SNAPSHOT_SYS_TIMESTAMP       = 1 << 0,
	SNAPSHOT_SYS_MEM_TOTALS      = 1 << 1,
	SNAPSHOT_SYS_MEM_USAGE       = 1 << 2,
	SNAPSHOT_SYS_MEM_WATERMARK   = 1 << 3,	/** non-standard feature */
	SNAPSHOT_SYS_CPU_MAX_FREQ    = 1 << 4,
	SNAPSHOT_SYS_CPU_USAGE       = 1 << 5,
	SNAPSHOT_SYS_CPU_FREQ        = 1 << 6,
	SNAPSHOT_SYS_MEM_CGROUPS     = 1 << 7,
	SNAPSHOT_SYS_MEM             = SNAPSHOT_SYS_MEM_TOTALS | SNAPSHOT_SYS_MEM_USAGE,
	SNAPSHOT_SYS_CPU             = SNAPSHOT_SYS_CPU_MAX_FREQ | SNAPSHOT_SYS_CPU_USAGE | SNAPSHOT_SYS_CPU_FREQ,
	SNAPSHOT_SYS                 = SNAPSHOT_SYS_TIMESTAMP | SNAPSHOT_SYS_CPU | SNAPSHOT_SYS_MEM
} sp_measure_sys_resource_t;


/**
 * Process resource identifiers.
 */
typedef enum {
	SNAPSHOT_PROC_MEM_USAGE      = 1 << 0,
	SNAPSHOT_PROC_CPU_USAGE      = 1 << 1,
	SNAPSHOT_PROC_MEM            = SNAPSHOT_PROC_MEM_USAGE,
	SNAPSHOT_PROC_CPU            = SNAPSHOT_PROC_CPU_USAGE,
	SNAPSHOT_PROC                = SNAPSHOT_PROC_MEM | SNAPSHOT_PROC_CPU
} sp_measure_proc_resource_t;


/**
 * Maemo5 specific - MEM watermark flags.
 */
typedef enum {
	MEM_WATERMARK_LOW    = 1 << 0,
	MEM_WATERMARK_HIGH   = 1 << 1
} sp_measure_mem_watermark_t;

#define ESPMEASURE_UNDEFINED		(-1)

#include <sp_measure_system.h>
#include <sp_measure_process.h>

#endif


src/measure_utils.c
/*
 * This file is a part of sp-measure library.
 *
 * Copyright (C) 2010 by Nokia Corporation
 *
 * Contact: Eero Tamminen <eero.tamminen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301  USA
 */

/* virtual file system root support */
char sp_measure_fs_root[1] = "";
char* sp_measure_virtual_fs_root = sp_measure_fs_root;


src/measure_utils.h
/*
 * This file is a part of sp-measure library.
 *
 * Copyright (C) 2010-2012 by Nokia Corporation
 *
 * Contact: Eero Tamminen <eero.tamminen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301  USA
 */
#ifndef MEASURE_UTILS_H
#define MEASURE_UTILS_H

#define ARRAY_ITEMS(arr) (sizeof(query)/sizeof(query[0]))

/**
 * Key/value pairs for file parsing.
 */
typedef struct parse_query_t {
	/* the key to look for */
	char* key;
	/* address where value should be written */
	int* value;
	/* true if the value was retrieved */
	bool ok;
} parse_query_t;


/* root of the /proc file system. */
extern char sp_measure_fs_root[];
extern char* sp_measure_virtual_fs_root;

#endif


examples/res-monitor.c
/*
 * This file is a part of sp-measure library.
 *
 * Copyright (C) 2010 by Nokia Corporation
 *
 * Contact: Eero Tamminen <eero.tamminen@nokia.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301  USA
 */

/**
 * A simple example demonstrating the usage of sp-measure library.
 * It monitors system (and process if pid is specified) resource usage.
 * The data is printed into console with 1 second interval until the
 * program is aborted with ctrl+c.
 *
 * Compile:
 * 1) example in source package
 *    gcc -I../src -L../src/.libs res-monitor.c -lspmeasure -o res-monitor
 * 2) example installed with libsp-measure-dev
 *    gcc -O -Wall res-monitor.c -lspmeasure -o res-monitor
 *
 * run:
 *    LD_LIBRARY_PATH=../src/.libs ./res-monitor [<pid>]
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include <sp_measure.h>


int main(int argc, char* argv[])
{
	int monitored_process = 0;
	if (argc > 1) {
		monitored_process = atoi(argv[1]);
	}

	sp_measure_sys_data_t sys_data[2];
	sp_measure_sys_data_t* sys_data1 = &sys_data[0];
	sp_measure_sys_data_t* sys_data2 = &sys_data[1];

	sp_measure_proc_data_t proc_data[2];
	sp_measure_proc_data_t* proc_data1 = &proc_data[0];
	sp_measure_proc_data_t* proc_data2 = &proc_data[1];

	/* initialize system snapshots */
	if (sp_measure_init_sys_data(sys_data1, SNAPSHOT_SYS, NULL) < 0) {
		fprintf(stderr, "Failed to initialize first system snapshot\n");
	}
	if (sp_measure_init_sys_data(sys_data2, 0,  sys_data1) < 0) {
		fprintf(stderr, "Failed to initialize second system snapshot\n");
	}

	/* Initialize process snapshots if necessary */
	if (monitored_process) {
		if (sp_measure_init_proc_data(proc_data1, monitored_process, SNAPSHOT_PROC, NULL) < 0) {
			fprintf(stderr, "Failed to initialize first process snapshot\n");
		}
		if (sp_measure_init_proc_data(proc_data2, 0, 0, proc_data1) < 0) {
			fprintf(stderr, "Failed to initialize second process snapshot\n");
		}
	}

	/* get the initial system snapshot */
	if (sp_measure_get_sys_data(sys_data1, SNAPSHOT_SYS, NULL) < 0) {
		fprintf(stderr, "Failed to get system snapshot\n");
	}
	/* get the initial process snapshot if necessary */
	if (monitored_process && sp_measure_get_proc_data(proc_data1, SNAPSHOT_PROC, NULL) < 0) {
		fprintf(stderr, "Failed to get process snapshot\n");
	}

	/* print report header */
	printf("System:");
	if (monitored_process) {
		printf("                        %d %s", FIELD_PROC_PID(proc_data1), FIELD_PROC_NAME(proc_data1));
	}
	printf("\n");
	printf("used mem: change:  cpu%%: freq:  ");
	if (monitored_process) {
		printf("clean:   dirty:  change:  cpu%%:");
	}
	printf("\n");

	/* loop until aborted */
	while (1) {
		int sys_mem_change, sys_cpu_usage, sys_cpu_avg_freq;
		/* get the next system snapshot */
		if (sp_measure_get_sys_data(sys_data2, SNAPSHOT_SYS, NULL) < 0) {
			fprintf(stderr, "Failed to get system snapshot\n");
		}
		/* calculate and print resource usage differences from the previous system snapshot */
		sp_measure_diff_sys_mem_used(sys_data1, sys_data2, &sys_mem_change);
		sp_measure_diff_sys_cpu_usage(sys_data1, sys_data2, &sys_cpu_usage);
		sp_measure_diff_sys_cpu_avg_freq(sys_data1, sys_data2, &sys_cpu_avg_freq);
		printf("%8d %+8d %5.1f%% %5d",  FIELD_SYS_MEM_USED(sys_data2), sys_mem_change, (float)sys_cpu_usage / 100, sys_cpu_avg_freq / 1000);

		if (monitored_process) {
			int sys_cpu_ticks, proc_cpu_ticks, proc_mem_change;
			/* get the next process snapshot */
			if (sp_measure_get_proc_data(proc_data2, SNAPSHOT_PROC, NULL) < 0) {
				fprintf(stderr, "Failed to get process snapshot\n");
				exit(-1);
			}
			/* calculate and print resource usage differences from the previous process snapshot */
			sp_measure_diff_sys_cpu_ticks(sys_data1, sys_data2, &sys_cpu_ticks);
			sp_measure_diff_proc_cpu_ticks(proc_data1, proc_data2, &proc_cpu_ticks);
			sp_measure_diff_proc_mem_private_dirty(proc_data1, proc_data2, &proc_mem_change);
			printf("%8d %8d %+8d %5.1f%%", FIELD_PROC_MEM_PRIVATE_CLEAN(proc_data2), FIELD_PROC_MEM_PRIV_DIRTY_SUM(proc_data2), proc_mem_change,
					 sys_cpu_ticks ? (float)proc_cpu_ticks * 100 / sys_cpu_ticks : 0);

		}
		printf("\n");

		/* swap system and process snapshot references - so that the last snapshot becomes previous
		 * snapshot and the next snapshot is written over the previous snapshot becoming
		 * the last snapshot */
		sp_measure_sys_data_t* sys_data_swap = sys_data1;
		sys_data1 = sys_data2;
		sys_data2 = sys_data_swap;

		sp_measure_proc_data_t* proc_data_swap = proc_data1;
		proc_data1 = proc_data2;
		proc_data2 = proc_data_swap;

		sleep(1);
	}

	/* releases resources allocated by snapshots.
	 * As the program is aborted, this is unnecessary.
	 * Still normally every initialized snapshot should be freed afterwards.  */
	sp_measure_free_sys_data(&sys_data[0]);
	sp_measure_free_sys_data(&sys_data[1]);

	sp_measure_free_proc_data(&proc_data[0]);
	sp_measure_free_proc_data(&proc_data[1]);

	return 0;
}


src/sp_measure_process.h
/*
 * This file is a part of sp-measure library.
 *
 * Copyright (C) 2010 by Nokia Corporation
 *
 * Contact: Eero Tamminen <eero.tamminen@nokia.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301  USA
 */
#ifndef SP_MEASURE_PROCESS_H
#define SP_MEASURE_PROCESS_H

/**
 * @file sp_measure_process.h
 * API for process resource usage snapshots.
 *
 * This file contains API providing access to usage statistics of such
 * process resources as CPU, memory etc.
 *
 * Short example (without any error checking):
 * @code
 *    // declare snapshot data structures
 *    sp_measure_proc_data_t data1, data2;
 *    // initialize the first snapshot structure for process with pid 1234
 *    sp_measure_init_proc_data(&data1, 1234, SNAPSHOT_PROC, NULL);
 *    // initialize the second snapshot structure
 *    sp_measure_init_proc_data(&data2, 0, 0, &data2);
 *    // take first snapshot
 *    sp_measure_get_proc_data(&data1, NULL);
 *    // do some actions
 *    ...
 *    // take the second snapshot
 *    sp_measure_get_proc_data(&data2, NULL);
 *    // print statistics:
 *    int value;
 *    printf("Resource usage of the process %d(%s)\n", FIELD_PROC_PID(&data1), FIELD_PROC_NAME(&data2));
 *    sp_measure_diff_proc_mem_private_dirty(&data1, &data2, &value);
 *    printf("\tprivate dirty memory: %+d\n", value);
 *    // release snapshot data structures
 *    sp_measure_free_proc_data(&data1);
 *    sp_measure_free_proc_data(&data2);
 *  @endcode
 *
 *  res-monitor.c.gz file in the package documentation dir
 *  is a working source example of the API usage.
 */
#ifdef __cplusplus
extern "C" {
#endif


/**
 * Common process information.
 */
typedef struct sp_measure_proc_common_t {
	/* pid */
	int pid;
	/* the process name */
	char* name;

	/* path of the /proc/<pid>/smaps file */
	char* proc_smaps_path;
	/* path of the /proc/<pid>/data file */
	char* proc_stat_path;

	/* process common data reference counter */
	int ref_count;
} sp_measure_proc_common_t;

/**
 * Process resource usage snapshot
 */
typedef struct sp_measure_proc_data_t {
	/* common process data, initialized at the beginning and not a subject to change */
	sp_measure_proc_common_t* common;

	/* the snapshot name */
	char* name;

	/* process memory statistics (from /proc/<pid>/smaps) */
	int mem_private_clean;
	int mem_private_dirty;
	int mem_swap;
	int mem_size;
	int mem_shared_clean;
	int mem_shared_dirty;
	int mem_pss;
	int mem_rss;
	int mem_referenced;

	/* system time ticks spent by process */
	int cpu_stime;
	/* user time ticks spent by process */
	int cpu_utime;

} sp_measure_proc_data_t;


/**
 * Initializes process snapshot data structure.
 *
 * If sample_data parameter is NULL the 'global' process parameters like
 * name are retrieved from system. Otherwise they are copied from the
 * sample_data structure.
 * The rest of parameters are reseted to zero values.
 * Afterwards the internal data structure data must be freed with
 * sp_measure_free_proc_data() function.
 * @param[out] new_data    the process snapshot data structure to initialize.
 * @param[in] pid          the process id (ignored if sample_data is given).
 * @param[in] resources    a flag specifying which initial process resource
 *                         statistics should be retrieved (ignored if
 *                         sample_data is given).
 * @param[in] sample_data  An already initialized process snapshot.
 * @return                 0  - success
 *                         >0 - only part of the initial process resource statistics
 *                              was retrieved. The returned value consists of the failed
 *                              resource identifiers (see sp_measure_proc_resource_t
 *                              enumeration).
 *                         <0 - unrecoverable error during process resource statistics
 *                              retrieval. For example process does not exist.
 */
int sp_measure_init_proc_data(
		sp_measure_proc_data_t* new_data,
		int pid,
		int resources,
		const sp_measure_proc_data_t* sample_data
		);


/**
 * Reinitializes process snapshot data structure.
 *
 * This function reinitializes 'global' parameters (like process name) of the process
 * snapshot.
 * @param[in] data   the process snapshot data structure.
 * @return           0 for success.
 */
int sp_measure_reinit_proc_data(
		sp_measure_proc_data_t* data
		);

/**
 * Releases process snapshot data structure.
 *
 * This function releases the resources allocated by sp_measure_init_proc_data() function.
 * @param[in] data   the process statistics data structure to free.
 * @return            0 for success.
 */
int sp_measure_free_proc_data(
		sp_measure_proc_data_t* data
		);

/**
 * Retrieves process resource usage snapshot.
 *
 * The data data structure must be initialized by sp_measure_init_proc_data() or
 * sp_measure_init_cloned_proc_data() functions.
 * @param[out] data      the process statistics data structure.
 * @param[in] resources  a flag specifying which process resource statistics
 *                       should be retrieved
 * @param[in] name       the snapshot name (optional). Can be NULL if snapshot naming
 *                       is not required.
 * @return               0  - success
 *                       >0 - only part of the process resource statistics was retrieved.
 *                            The returned value consists of the failed resource
 *                            identifiers (see sp_measure_proc_resource_t enumeration).
 *                       <0 - unrecoverable error during process resource statistics
 *                            retrieval.
 */
int sp_measure_get_proc_data(
		sp_measure_proc_data_t* data,
		int resources,
		const char* name
		);


/**
 * The process snapshot field comparison functions.
 *
 * All process snapshot field comparison functions have the following signature:
 * int (*)(const sp_measure_proc_data_t* data1, const sp_measure_proc_data_t* data2, int* diff);
 *
 * @param[in] data1    the first snapshot.
 * @param[in] data2    the second snapshot.
 * @param[out] diff    the difference of the field value between snapshots.
 * @return             0 if the difference was calculated successfully.
 */


/**
 * Retrieves process private dirty memory difference (in KB) between two snapshots.
 *
 * @param[in] data1  the first snapshot.
 * @param[in] data2  the second snapshot.
 * @param[out] diff  the memory difference.
 * @return           0 for success.
 */
int sp_measure_diff_proc_mem_private_dirty(
		const sp_measure_proc_data_t* data1,
		const sp_measure_proc_data_t* data2,
		int* diff
		);

/**
 * Retrieves time spent in process between two snapshots.
 *
 * @param[in] data1  the first snapshot.
 * @param[in] data2  the second snapshot.
 * @param[out] diff  the number of cpu ticks spent in process.
 * @return           0 for success.
 */
int sp_measure_diff_proc_cpu_ticks(
		const sp_measure_proc_data_t* data1,
		const sp_measure_proc_data_t* data2,
		int* diff
		);

/*
 * Field access definitions
 */

#define FIELD_PROC_PID(data)                 (data)->common->pid
#define FIELD_PROC_NAME(data)                (data)->common->name
#define FIELD_PROC_MEM_PRIVATE_CLEAN(data)   (data)->mem_private_clean
#define FIELD_PROC_MEM_PRIVATE_DIRTY(data)   (data)->mem_private_dirty
#define FIELD_PROC_MEM_SWAP(data)            (data)->mem_swap
#define FIELD_PROC_MEM_SIZE(data)            (data)->mem_size
#define FIELD_PROC_MEM_PSS(data)             (data)->mem_pss
#define FIELD_PROC_MEM_RSS(data)             (data)->mem_rss
#define FIELD_PROC_MEM_REFERENCED(data)      (data)->mem_referenced
#define FIELD_PROC_MEM_SHARED_CLEAN(data)    (data)->mem_shared_clean
#define FIELD_PROC_MEM_SHARED_DIRTY(data)    (data)->mem_shared_dirty
#define FIELD_PROC_MEM_PRIV_DIRTY_SUM(data)  (FIELD_PROC_MEM_SWAP(data) + FIELD_PROC_MEM_PRIVATE_DIRTY(data))

#define FIELD_PROC_CPU_STIME(data)           (data)->cpu_stime
#define FIELD_PROC_CPU_UTIME(data)           (data)->cpu_utime

#ifdef __cplusplus
}
#endif

#endif
